delphi/translators/for2py/genPGM.py

```diff
@@ -61,409 +67,763 @@
+        return list(
+            chain.from_iterable(
+                [
+                    self.gen_grfn(cur, state, call_source)
+                    for cur in node
+                ]
+            )
+        )
 
-            body, fns, iden_spec = get_body_and_functions(bodyPgm)
+    def process_function_definition(self, node, state, *_):
+        """
+            This function processes the function definition i.e. functionDef
+            instance. It appends GrFN dictionaries to the `functions` key in
+            the main GrFN JSON. These dictionaries consist of the
+            function_assign_grfn of the function body and the
+            function_container_grfn of the function. Every call to this
+            function adds these along with the identifier_spec_grfn to the
+            main GrFN JSON.
+        """
 
-            variables = list(localDefs.keys())
-            variables_tmp = []
+        # Add the function name to the list that stores all the functions
+        # defined in the program
+        self.function_definitions.append(node.name)
 
-            for item in variables:
-                match = re.match(
-                    r"(format_\d+_obj)|(file_\d+)|(write_list_\d+)|"
-                    r"(write_line)",
-                    item,
-                )
-                if not match:
-                    variables_tmp.append(item)
+        local_last_definitions = state.last_definitions.copy()
+        local_next_definitions = state.next_definitions.copy()
+        local_variable_types = state.variable_types.copy()
+        scope_path = state.scope_path.copy()
 
-            variables = variables_tmp
+        # If the scope_path is empty, add _TOP to the list to denote that
+        # this is the outermost scope
+        if len(scope_path) == 0:
+            scope_path.append("_TOP")
+        scope_path.append(node.name)
+
+        if node.decorator_list:
+            # This is still a work-in-progress function since a complete
+            # representation of SAVEd variables has not been decided for GrFN.
+            # Currently, if the decorator function is static_vars (for
+            # SAVEd variables), their types are loaded in the variable_types
+            # dictionary.
+            function_state = state.copy(
+                last_definitions=local_last_definitions,
+                next_definitions=local_next_definitions,
+                function_name=node.name,
+                variable_types=local_variable_types,
+            )
+            process_decorators(node.decorator_list, function_state)
+
+        # Check if the function contains arguments or not. This determines
+        # whether the function is the outermost scope (does not contain
+        # arguments) or it is not (contains arguments). For non-outermost
+        # scopes, indexing starts from -1. For outermost scopes, indexing
+        # starts from 0
+        # TODO: What do you do when a non-outermost scope function does not
```

Comment:
@pratikbhd Is it possible to add an `assert` somewhere in for2py (not necessarily in gen_PGM.py, but wherever it makes sense) to catch this situation?  That way we'll easily know what the problem is if this situation every crops up.